accepted/2021/statics-in-interfaces/README.md

```diff
@@ -667,21 +671,30 @@ The numeric interfaces build upon the base interfaces by defining the core abstr
 ```csharp
 namespace System
 {
-    public interface INumber<TSelf>
+    public interface INumberBase<TSelf>
         : IAdditionOperators<TSelf, TSelf, TSelf>,
           IAdditiveIdentity<TSelf, TSelf>,
-          IComparisonOperators<TSelf, TSelf>,   // implies IEqualityOperators<TSelf, TSelf>
+          IEqualityOperators<TSelf, TSelf>,     // implies IEquatable<TSelf>
+          IMultiplyOperators<TSelf, TSelf>,
+          ISubtractionOperators<TSelf, TSelf, TSelf>,
+          IUnaryPlusOperators<TSelf, TSelf>
+        where TSelf : INumberBase<TSelf>
+    {
+        // Alias for AdditiveIdentity
+        static abstract TSelf Zero { get; }
+    }
```

Comment:
For right now, this effectively mirrors the `Numerics` protocol in `Swift`. Noting that the missing bits would be:
* `TMagnitude Magnitude { get; }` as there are no associated types and so `TSelf Abs(TSelf value)` is the closest alternative (https://developer.apple.com/documentation/swift/numeric/2884876-magnitude)
* `TSelf init(IntegerLiteralType)`, again due to no associated types (https://developer.apple.com/documentation/swift/expressiblebyintegerliteral)
* `TSelf? init(TBinaryInteger)`, as there are some open questions here that were raised on #205 (https://developer.apple.com/documentation/swift/numeric/2886795-init)

Reply:
Going to call out each of the remaining interfaces/members on `INumber` and discuss why they exist. People should weigh in on why they shouldn't be on `INumberBase` (name is a placeholder atm).

First lets cover the ones that make sense to leave off:
* `IComparisonOperators` - This one actually makes sense to leave off as it can't be supported by things like `Complex`
* `IModulusOperators` - This one probably makes sense to leave off as it doesn't make a lot of sense for `Complex`

The ones that probably make sense to have:
* `IDecrementOperators` - This is, for many reasons, simply `x - 1`. It is a general nicety in .NET and something often overlooked
* `IDivisionOperators` - Complex numbers support division. What scenarios are people wanting where division isn't possible but something is still a "number"
* `IIncrementOperators` - This is, for many reasons, simply `x + 1`. It is a general nicety in .NET and something often overlooked
* `IMultiplicativeIdentity` - Complex numbers have a multiplicative identity and it is `1`
* `IUnaryNegationOperators` - This is logically `Zero - x` and `INumberBase` supports subtraction, zero, and the additive identity

The "controversial" ones:
* `ISpanFormattable` - All types support `ToString`. Supporting culture aware formatting of your data is a "best practice" and helps support globally aware apps. In the worst case, you implement this as `=> ToString();`
* `ISpanParseable` - Parsing can get complex in some cases but it is likewise a "best practice" for number types to support something built in. This helps support all kinds of programs, roundtripping of the format strings, and both culture aware/invariant data processing.

Reply:
There are also some individual APIs:
* `Abs` - Potentially makes sense, would be "inefficient" for types like `Complex`
* `DimRem` - Doesn't make sense without `IModulusOperators`
* `Max`, `Min`, and `Clamp` - Doesn't make sense for types without `IComparisonOperators`
* `Sign` - May or may not make sense, it can get complicated when dealing with things like `Complex` on what the sign is

Also there is: `Create`, `CreateSaturating`, and `CreateTruncating`. As on #205 these exist because there are a lot of scenarios that get cut off when you can't convert from a `T` to a `U`. For example, you can't correctly account for or handle overflow or take in user-defined values.

You'd have to specialize on each and every `T` you want to support (keeping in mind overloading on generic constraints isn't possible) and ultimately throw anyways if you can't get the data you need back out. For something like `Swift`, you can only create a number from a `BinaryInteger` and so you can't for example go from `float->double` or `float->int`, etc.

Reply:
`Abs` is actually the same as `Magnitude`. It makes sense for `Complex`, but returns a `double` (which ought to be be an associated type). It _can_ be made to return Complex for complex as well (akin to the function `Trunc` which returns `double` and not `int` for `double` argument).
`DivRem` is usually considered only for discrete types, but can be naturally extended to doubles (`div = Math.Trunc(a / b); rem = a - b * div`) and even to Complex using the same computation (`Trunc` on a Complex should internally truncate both components, bringing the original Complex to the 1x1 grid).
`Sign` is actually `c / |c|` = a complex with the same [argument](https://en.wikipedia.org/wiki/Complex_number#Polar_complex_plane) and magnitude of 1.

Reply:
> Abs is actually the same as Magnitude

In the strictest sense of the definition of "magnitude" vs "absolute value", yes. In terms of how I was referring to them above, no.

`Abs` in .NET returns the same type as is input and so you have `int Abs(int)` and it may fail for `int.MinValue`. Where-as `Magnitude` in swift returns an associated type which is unsigned for two's complement values and so it would be `uint Magnitude { get; }`. We don't have associated types and will most likely not be getting them before generic math officially releases so they should be considered not a possible solution here.

> I would not include Abs into the general INumber or INumberBase interface, because there are useful number abstractions (like integers modulo n which cannot define Magnitude in a useful way). If we would include Abs into INumber, the users would be unable to represent such structures as an INumber.

.NET integers already default to two's complement modular arithmetic. They wrap around on overflow by default and so `int.MaxValue + 1 == int.MinValue`. I don't believe its worth cutting off absolute value to support increasingly esoteric use cases and instead such scenarios can decide upon a correct behavior for that scenario instead (e.g. `Abs(int.MinValue)` will throw; `Abs(uint)` just returns itself, etc).

Reply:
> .NET integers already default to two's complement modular arithmetic. They wrap around on overflow by default and so int.MaxValue + 1 == int.MinValue. I don't believe its worth cutting off absolute value to support increasingly esoteric use cases and instead such scenarios can decide upon a correct behavior for that scenario instead (e.g. Abs(int.MinValue) will throw; Abs(uint) just returns itself, etc).

Well, `Abs` is already pointless not only on esoteric types, but actually on more mundane types like unsigned everything. Okay, esoteric types can just implement fine-grained interfaces and not the generic `INumberBase`, which solves the problem. But having `Abs` on unsigned types feels impure.

From the other point of view, if `Abs` returns the same type as the argument, `Complex` case is easy (we just return the mathematical magnitude), which is a `Complex` itself).

Reply:
> But having Abs on unsigned types feels impure.

Logically such an operation makes sense and it is simply a `no-op`, same as it would be for any signed input that is already positive.

In the context of concrete types, you can manually avoid the operation knowing that its always positive.

In the context of generic math, unless you explicitly constrain to `ISignedNumber`, then you don't know if the number you have could be negative. You can't overload on generic constraint and duplicating tons of logic just to differ on `Abs` isn't an ideal setup anyways; nor is explicitly checking if a type could be negative. Instead, just expose the operation, have it do nothing (just like `operator +(TSelf value)`) and let the JIT or other compiler correctly optimize it.